### DISCUSSION TITLE
STR-DN1080 - Some APIs work, some return "illegal argument (3)"

### DISCUSSION THREAD

**Discussion by SillaRizzoli** [jensrossbach/node-red-contrib-sony-audio#34] [Q&A]
Hello,
I'm trying to use this wonderful node to control my fully updated Sony STR-DN1080.
The problem is that some features work, like volume up/down, changing the input, but many other return "illegal argument (3)".
I've tried using postman to investigate the issue and, sure enough, I can easily control the features that already work with the node, but all the others return an error similar to the following:
{
    "error": [
        12,
        "getSupportedApiInfo"
    ],
    "id": 5
}
I'm trying examples right from the developer.sony.com website, so I'm pretty sure that I'm not messing anything up. The APIs I'm calling are listed as supported for my home theater.

Am I missing something?

Thanks for your attention!

Best regards
Silla

**Comment by jensrossbach** [comment]
Hi,
thank you that you like my Node-RED nodes.

I'm not familiar with postman but I assume that you can access web APIs with this tool? If the same commands that do not work with the Sony nodes also do not work by directly executing with postman, I would assume that the problem is at the Sony STR-DN1080 device itself. Did you recently update the firmware which might cause these issues? Did the problematic commands work in the past?

I don't have a Sony STR-DN1080, so I cannot test this myself but nevertheless maybe you can post some examples of API commands which do not work (i.e., full JSON spec with all applied parameters)?

Best regards
Jens

**Comment by SillaRizzoli** [comment]
Hi,
I've been investigating a little more in the meantime, and half of what I've written above is not actually correct; the error 12 (No Such Method) I was getting is because I was querying the wrong URL /sony/system instead of /sony/audio. However, your node still send queries that result in error 3 (illegal argument); this happens with the "Controller" node, "Set Sound Settings" command and all the sub-parameters.
I've used wireshark to intercept the commands sent by your node and they look like this:

#Nightmode switched on
{
	"id": 1,
	"method": "setSoundSettings",
	"version": "1.1",
	"params": [
		{
			"settings": [
				{
					"target": "nightMode",
					"value": "on"
				}
			]
		}
	]
}

#Clearaudio {
  "id": 1,
  "method": "setSoundSettings",
  "version": "1.1",
  "params": [
    {
      "settings": [
        {
          "target": "soundField",
          "value": "clearAudio"
        }
      ]
    }
  ]
}

I've replayed them with postman and I get the same error, so I dug a little deeper: I managed to execute the getSoundSettings command and the result is below:
#Query:
{"id":1,"method":"getSoundSettings","version":"1.1","params":[{"settings":[{"target":""}]}]}

#Results:
{
    "id": 1,
    "result": [
        [
            {
                "candidate": [
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "2ch Stereo",
                        "titleTextID": "sound-soundfield-afd-2ch",
                        "value": "2chStereo"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Multi Ch Stereo",
                        "titleTextID": "sound-soundfield-multichstereo",
                        "value": "multiChStereo"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Direct",
                        "titleTextID": "sound-soundfield-multi-direct",
                        "value": "direct"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "A.F.D.",
                        "titleTextID": "sound-soundfield-audioformatdecoding",
                        "value": "audioFormatDecoding"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Dolby Surround",
                        "titleTextID": "sound-soundfield-dolbysurround",
                        "value": "dolbySurround"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Neural:X",
                        "titleTextID": "sound-soundfield-neuralx",
                        "value": "neuralX"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Front Surround",
                        "titleTextID": "sound-soundfield-movie-frontsurround",
                        "value": "frontSurround"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Audio Enhancer",
                        "titleTextID": "sound-soundfield-music-audioenhancer",
                        "value": "audioEnhancer"
                    },
                    {
                        "isAvailable": false,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "",
                        "titleTextID": "sound-soundfield-off",
                        "value": "off"
                    }
                ],
                "currentValue": "neuralX",
                "deviceUIInfo": "soundFieldFig",
                "isAvailable": true,
                "target": "soundField",
                "title": "Sound Field",
                "titleTextID": "sound-soundfield",
                "type": "enumTarget"
            },
            {
                "candidate": [
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "On",
                        "value": "on"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Off",
                        "value": "off"
                    }
                ],
                "currentValue": "off",
                "isAvailable": true,
                "target": "pureDirect",
                "title": "Pure Direct",
                "titleTextID": "sound-puredirect",
                "type": "booleanTarget"
            },
            {
                "candidate": [
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Off",
                        "titleTextID": "sound-soundoptimizer-off",
                        "value": "off"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Normal",
                        "titleTextID": "sound-soundoptimizer-normal",
                        "value": "normal"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Low",
                        "titleTextID": "sound-soundoptimizer-normal",
                        "value": "low"
                    },
                    {
                        "isAvailable": false,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "",
                        "titleTextID": "sound-soundoptimizer-off",
                        "value": ""
                    }
                ],
                "currentValue": "normal",
                "isAvailable": true,
                "target": "optimizer",
                "title": "Sound Optimizer",
                "titleTextID": "sound-soundoptimizer",
                "type": "enumTarget"
            },
            {
                "candidate": [
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Off",
                        "titleTextID": "sound-calibrationtype-off",
                        "value": "off"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Full Flat",
                        "titleTextID": "sound-calibrationtype-fullflat",
                        "value": "fullFlat"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Engineer",
                        "titleTextID": "sound-calibrationtype-engineer",
                        "value": "engineer"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Front Reference",
                        "titleTextID": "sound-calibrationtype-frontreference",
                        "value": "frontReference"
                    },
                    {
                        "isAvailable": false,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "",
                        "titleTextID": "",
                        "value": ""
                    }
                ],
                "currentValue": "engineer",
                "isAvailable": true,
                "target": "calibrationType",
                "title": "Calibration Type",
                "titleTextID": "sound-calibrationtype",
                "type": "enumTarget"
            },
            {
                "candidate": [
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "1: Movie ",
                        "titleTextID": "custompreset-custompreset-movie",
                        "value": "movie"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "2: Music ",
                        "titleTextID": "custompreset-custompreset-music",
                        "value": "music"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "3: Party ",
                        "titleTextID": "custompreset-custompreset-party",
                        "value": "party"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "4: Night ",
                        "titleTextID": "custompreset-custompreset-night",
                        "value": "night"
                    },
                    {
                        "isAvailable": true,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "Undo",
                        "titleTextID": "custompreset-custompreset-off",
                        "value": "off"
                    },
                    {
                        "isAvailable": false,
                        "max": -1,
                        "min": -1,
                        "step": -1,
                        "title": "",
                        "titleTextID": "custompreset-custompreset-null",
                        "value": ""
                    }
                ],
                "currentValue": "",
                "deviceUIInfo": "soundFieldFig",
                "isAvailable": true,
                "target": "sceneSelection",
                "title": "Custom Preset",
                "titleTextID": "custompreset-custompreset",
                "type": "enumTarget"
            }
        ]
    ]
}

Now what can be done is much clearer: there are five settings that can be modified:
#This is essentially the surround mode
0: soundField
    parameters: 2chStereo, multiChStereo, direct, audioFormatDecoding, dolbySurround, neuralX, frontSurround, audioEnhancer, off

#This enables/disables every modification to the input signal (and turns off/on the front LCD.
1: pureDirect
    parameters: on, off

#This modifies the dynamic range
2: optimizer
    parameters: off, normal, low

#This modifies the calibration
3:  calibrationType
    parameters: off, fullFlat, engineer, frontReference

#This allows to choose from different presets that can be modified with the remote
4: sceneSelection
    parameters: movie, music, party, night, off

If I send, for example, {"id":1,"method":"setSoundSettings","version":"1.1","params":[{"settings":[{"target":"soundField","value":"neuralX"}]}]} to the device it works, just like any other combination above.
Unfortunately, it looks like there's no way to modify individual parameters like nightMode from the API, , even tough the option is there with the remote. :(

So, you were right that the problem lies with my device; I'll try to modify the custom presets to see if I can adapt them to my needs, it's easy enough to activate them with an http request node.

If you want to add specific code for the STR-DN1080 let me know, I'll be more than happy to help you by sending all the JSONs you need and do the testing, but I don't have much experience with javascript or html.

Thank you very much for your help!

**Reply by jensrossbach** [reply]
Good to know that you are more successful now. I based the development of the Sony nodes purely on the public API documentation and what I was able to test with my Sony HT-XT3. As I don't have any other Sony audio device, I cannot test specific features of these other devices. But for sure I'm willing to add more features with the help from others having these devices. So if you can send me such information as you proposed and later support me in testing, I can extend the nodes accordingly.
